docs/derived-variables.qmd

```diff
@@ -0,0 +1,256 @@
+enables the execution of arbitrary Python code within a .R file. Example below,
+
+```{r, python-in-R}
+# Python code that creates a function to calculate BMI
+BMI_python_code <- "def BMI(weight, height): 
+  return weight/(height*height)
+" 
+# Run reticulate to bring the Python function over to the R environment enabling
+# us to call it using R code
+reticulate_output <- reticulate::py_run_string(BMI_python_code)
+# Calculate BMI using the created Python function
+BMI_python <- reticulate_output$BMI(25, 5)
+# Expect the calculated BMI to be 1
+stopifnot(BMI_python == 1)
+print(paste("Calculated BMI:", BMI_python))
+```
+
+To reduce the library scope, currently only R functions will be allowed.
+
+### Function Inputs and Outputs
+
+The function inputs and output should be standardized to accept either scalar or
+vector values.
+
+For example, a function that calculates BMI will take as input height and
+weight. The library should make a decision on whether the inputs can be scalar
+values for example 80 and 180 or vector values for example c(80) and c(180).
+
+This decision will have implications on the performance, user experience, and
+portability of the function as described below.
+
+R being a statistical language is optimized to run on vectors. However, a
+function accepting scalar inputs can be easily vectorized using the
+[Vectorize](https://stat.ethz.ch/R-manual/R-devel/library/base/html/Vectorize.html)
+function available in base R.
+
+The code below displays a table that compares the performance of calculating BMI
+using:
+
+1.  Scalar inputs
+2.  Vector inputs
+3.  A vectorized version of 1 created using the `Vectorize` function
+
+```{r}
+scalar_BMI <- function(weight, height) {
+  bmi <- NA
+  if(is.na(weight) | is.na(height)) {
+    bmi <- NA
+  }
+  bmi <- weight/(height*height)
+}
+
+vector_BMI <- function(weight, height) {
+    bmi <- ifelse(
+      is.na(weight) | is.na(height),
+      NA,
+      weight/(height*height)
+    )
+}
+
+vectorized_BMI <- Vectorize(scalar_BMI)
+
+# Create larger test dataset for better comparison
+test_data <- data.frame(
+  weight = sample(1:100, 1000, replace = TRUE),
+  height = sample(1:100, 1000, replace = TRUE)
+)
+
+# In milliseconds
+get_func_perf <- function(func) {
+  start_time <- Sys.time()
+  func()
+  end_time <- Sys.time()
+  run_time <- end_time - start_time
+  return(as.numeric(run_time)*1000)
+}
+
+# Original methods
+scalar_run_time <- get_func_perf(function() {
+  bmi <- c()
+  for(i in 1:nrow(test_data)) {
+    bmi <- c(bmi, scalar_BMI(test_data[i, "weight"], test_data[i, "height"]))
+  }
+})
+
+vector_run_time <- get_func_perf(function() {
+  vector_BMI(test_data$weight, test_data$height)
+})
+
+vectorized_run_time <- get_func_perf(function() {
+  vectorized_BMI(test_data$weight, test_data$height)
+})
+
+# Mapply methods - testing all three functions
+mapply_vectorized_run_time <- get_func_perf(function() {
```

Comment:
@DougManuel Does it makes sense to run the mapply tests with the vector BMI functions? It will essentially run them as a scalar function passing in each value in the vector one at a time, defeating the purpose of those functions accepting a vector.